### DISCUSSION TITLE
OMI Identity on Element Matrix

### DISCUSSION THREAD

**Discussion by fire** [omigroup/omigroup#77] [R&D]
Discuss and propose using Matrix Element for identity.

**Comment by fire** [comment]
Press to open :point_down: 
<details><summary>Log 2021-08-19 on Identity on Element Matrix</summary>
<p>

iFire — Yesterday at 13:31
@Robert for funsie you saw there's two matrix client libraries right?
Robert — Yesterday at 13:31
There's actually quite a few
iFire — Yesterday at 13:31
https://github.com/Sorunome/miitrix
there's a second one
Koolala — Yesterday at 13:32
I actually just now fell in love with ActivityPub since I read its all JSON
Like a GLTF would possibly just work with it automatically cause its extensions are basically: is it described in JSON
iFire — Yesterday at 13:33
https://github.com/Nheko-Reborn/mtxclient
so both are the free-est extent of remixing. MIT or similar
Robert — Yesterday at 13:34
ActivityPub and Matrix are both all JSON
I'd take a look at the spec
https://matrix.org/docs/spec/client_server/r0.6.1
Or here's one that lets you interact with a server via curl:
https://matrix.org/docs/guides/client-server-api
Matrix.org
Client Server API | Matrix.org
How to use the client-server API If you haven't already, get a homeserver up and running on
https://localhost:8448 - e.g. by following the…

You don't really need a client SDK, you could roll your own. It's a bit of work to get everything working like E2E encryption and message syncing, but it's doable.
It's all HTTP based.
ActivityPub is as well.
The real differences are in the intended usecases.
iFire — Yesterday at 13:38
Is it hard to support enough to get Robert's name and his avatar binary blob from an identfier?
Robert — Yesterday at 13:39
ActivityPub was basically designed to be an federated alternative to social feeds like Twitter. Matrix was designed to be a federated chat/voip protocol.
Robert — Yesterday at 13:39
Shouldn't be too hard
https://matrix.org/docs/spec/#profiles
https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-profile-userid
iFire — Yesterday at 13:41
You also mentioned a mailbox on your identity spec and @humbletim developed the concept of a friend is only when both contacts friend each other.
Robert — Yesterday at 13:42
Yeah both Matrix and ActivityPub don't have a concrete method of friending eachother.
iFire — Yesterday at 13:43
there's also merge into a group which a matrix room gives
Robert — Yesterday at 13:43
Like you said, for activity pub, it's a mutual following. For Matrix, you just get the list of 1-1 rooms. You have to accept room invites anyway. It's not really the same as a friend. I guess you could set room information to say that this is a friend room though.
And yeah you can represent groups as rooms as well.
iFire — Yesterday at 13:44
There's the Japanese cultural concept of artist circles.
Robert — Yesterday at 13:44
And then to notfiy a user/group you'd send it to all of their mailboxes in activity pub or the room in matrix.
iFire — Yesterday at 13:45
and each object is signed by a retractable signature! by default
Robert — Yesterday at 13:45
Yup
they could even be encrypted in matrix
ActivityPub may have a standard for that as well idk
iFire — Yesterday at 13:48
the object signing should be acceptable by the blockchain crypto developers
for use in whatever tech they want to use
matrix signing for clarity is not on the blockchain
which is what we promised we would do to not require blockchain like Eth or Bitcoin
Koolala — Yesterday at 14:09
I wouldn't really be worried about intention or if its named a chat room or a feed or a message or a post but I'm curious to see what the overlap is with matrix and actpub just in terms of identity and organized communication over time
iFire — Yesterday at 14:15
@Robert does matrix has a webfinger http json equiv?
Robert — Yesterday at 14:18
In what sense? Email -> MatrixID?
https://matrix.org/docs/spec/identity_service/r0.3.0#email-associations
iFire — Yesterday at 14:19
url/MatrixID -> Name, GLB, Mailbox
Robert — Yesterday at 14:19
That'd be this: https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-profile-userid
Server -> Server version: https://matrix.org/docs/spec/server_server/r0.1.4#get-matrix-federation-v1-query-profile
iFire — Yesterday at 14:23
so in theory you can send me a matrix client http url for robert?
Robert — Yesterday at 14:24
https://matrix.org/_matrix/client/r0/profile/@arobertlong:matrix.org
iFire — Yesterday at 14:25
{"displayname":"Robert Long","avatar_url":"mxc://matrix.org/YsMmlvGvDByojlKdmOVEcnGT"}
Robert — Yesterday at 14:25
Yup
That's it
Now if I can add my glb avatar to it, then we're good to go
I think that could be added to the spec and Synapse for sure
iFire — Yesterday at 14:26
there's private data section
you linked to it
3.7   Profiles

Users may publish arbitrary key/value data associated with their account - such as a human readable display name, a profile photo URL, contact information (email address, phone numbers, website URLs etc).

3.8   Private User Data

Users may also store arbitrary private key/value data in their account - such as client preferences, or server configuration settings which lack any other dedicated API. The API is symmetrical to managing Profile data.
Robert — Yesterday at 14:27
Yeah, I'd need to look into that more. I haven't looked closely yet.
Koolala — Yesterday at 14:29
What parts of activitypub are you for sure confident of in their ability to overlap with Matrix? @Robert
In a standard way
Robert — Yesterday at 14:31
They overlap in that they both can sync federated lists of events. Events which can be extended in similar ways. They both provide identity, but Matrix also specifies possible authentication models.
iFire — Yesterday at 14:32
Q? is the avatar binary blob completely open like the open web, or do each platform need to get it via capabilities. which requires rooms
Robert — Yesterday at 14:32
Kinda up to us to figure that out
I could see either.
Idk, lots of stuff to figure out.


</p>
</details>

**Reply by fire** [reply]
Press to open :point_down: 

<details><summary>Log 2021-08-19 on Element Matrix client MVP</summary>
<p>

1. Authenticate with matrix
2. Join a room tagged as a 3D world
3. Start V-Sekai Servers and offer webrtc channels
4. Start V-Sekai clients via offered webrtc datachannel

How do I do this in a smooth timescale?

I think the link I found would be the quickest
https://github.com/Nheko-Reborn/mtxclient

Robert — Today at 17:37

I'll have a read through

Yeah this looks pretty good :thumbsup:

https://github.com/Nheko-Reborn/mtxclient/blob/master/examples/room_feed.cpp
GitHub

mtxclient/room_feed.cpp at master · Nheko-Reborn/mtxclient
Client API library for Matrix, built on top of Boost.Asio - mtxclient/room_feed.cpp at master · Nheko-Reborn/mtxclient

That's pretty concise

Ok, I gotta go, but I have a lot to talk about how you might specify the different layers of the protocol and could use feedback on those.

iFire — Today at 17:44

not sure, the first parts are already some work

will design those other layers later

Robert — Today at 17:45

Looking at that library, I think it'll be a pretty achievable MVP.
</p>
</details>

**Comment by robertlong** [comment]
Thanks for posting this! Soon I'd like to get a more formal proposal together so we can talk about how platforms might utilize Matrix to build out realtime communications, identity, and permission systems.
